conda/core/index.py

```diff
@@ -51,7 +58,457 @@ def check_allowlist(channel_urls: list[str]) -> None:
 LAST_CHANNEL_URLS = []
 
 
+class Index(UserDict):
+    """The ``Index`` provides information about available packages from all relevant sources.
+
+    There are four types of sources for package information, namely
+
+    Channels
+        represent packages available from standard sources identified with a url, mostly online,
+        but can also be on a local filesystem using the `file://` scheme.
+        Programatically, channels are represented by :class:`conda.models.Channel`, their data
+        is fetched using :class:`conda.index.subdir_data.SubdirData`.
+
+        For more information see :ref:`condepts-channels`.
+
+    Prefix
+        represents packages that are already installed. Every :class:`Index` can be associated
+        with exactly one Prefix, which is the location of one of the conda :ref:`concepts-environments`.
+        The package information about the installed packages is represented by :class:`conda.prefix.prefix_data.PrefixData`.
+
+    Package Cache
+        represents packages that are locally unpacked, but may not be installed in the environment
+        associated with this index. These are usually packages that have been installed in any environment
+        of the local conda installation, but may have been removed from all environments by now.
+
+    Virtual Packages
+        represent properties of the system, not actual conda packages in the normal sense. These are,
+        for example, system packages that inform the solver about the operating system in use, or
+        track features that can be used to steer package priority.
+    """
+
+    def __init__(
+        self,
+        channels=(),
+        prepend=True,
+        platform=None,
+        subdirs=None,
+        use_local=False,
+        use_cache=None,
+        prefix=None,
+        repodata_fn=context.repodata_fns[-1],
+        add_system=False,
+    ) -> None:
+        if use_local:
+            channels = ["local"] + list(channels)
+        if prepend:
+            channels += context.channels
+        self._channels = channels
+        if subdirs:
+            if platform:
+                log.warning("subdirs is %s, ignoring platform %s", subdirs, platform)
+        else:
+            subdirs = (platform, "noarch") if platform is not None else context.subdirs
+        self._subdirs = subdirs
+        self._repodata_fn = repodata_fn
+        self.channels = {}
+        self.expanded_channels = []
+        for channel in channels:
+            urls = Channel(channel).urls(True, subdirs)
+            check_allowlist(urls)
+            expanded_channels = [Channel(url) for url in urls]
+            self.channels[channel] = [
+                SubdirData(expanded_channel, repodata_fn=repodata_fn)
+                for expanded_channel in expanded_channels
+            ]
+            self.expanded_channels.extend(expanded_channels)
+        LAST_CHANNEL_URLS.clear()
+        LAST_CHANNEL_URLS.extend(self.expanded_channels)
```

Comment:
Can you add a note about this global state bit and why is it needed? We should aim to remove it at some point.